src/commandbar.js

```diff
@@ -0,0 +1,268 @@
+                    this.show();
+                    event.target.removeEventListener('keyup', keyUp, true);
+                }
+            };
+            event.target.addEventListener('keyup', keyUp, true);
+            this._active = true;
+            this.render(this.options.commands);
+            this._resetPosition();
+
+            let initialTarget;
+            let oldValue;
+
+            const keyup = event => {
+                if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+                    event.preventDefault();
+                    return;
+                }
+                if (event.key === '/' && !initialTarget) {
+                    initialTarget = event.target;
+                    oldValue = event.target.innerText;
+                }
+                const diff = patienceDiff(
+                    oldValue.split(''),
+                    initialTarget.innerText.split(''),
+                    true,
+                );
+                this._lastText = diff.bMove.join('').trim();
+
+                if (this._lastText.match(/\s/)) {
+                    this._stop();
+                    return;
+                }
+                const term = this._lastText;
+
+                this._currentFilteredCommands = this._filter(term);
+                this.render(this._currentFilteredCommands);
+                this._resetPosition();
+            };
+            const keydown = e => {
+                if (e.key === 'Enter') {
+                    e.stopImmediatePropagation();
+                    this._currentValidate();
+                    e.preventDefault();
+                } else if (e.key === 'Escape') {
+                    e.stopImmediatePropagation();
+                    this._stop();
+                    e.preventDefault();
+                } else if (e.key === 'Backspace' && !this._lastText) {
+                    this._stop();
+                } else if (e.key === 'ArrowDown' || e.key === 'ArrowUp') {
+                    e.preventDefault();
+                    e.stopImmediatePropagation();
+
+                    const index = this._currentFilteredCommands.findIndex(
+                        c => c === this._currentSelectedCommand,
+                    );
+                    if (!this._currentFilteredCommands.length || index === -1) {
+                        this._currentSelectedCommand = undefined;
+                    } else {
+                        const n = e.key === 'ArrowDown' ? 1 : -1;
+                        const newIndex = cycle(index + n, this._currentFilteredCommands.length - 1);
+                        this._currentSelectedCommand = this._currentFilteredCommands[newIndex];
+                    }
+                    e.preventDefault();
+                    this.render(this._currentFilteredCommands);
+                }
+            };
+            const mousemove = () => {
+                this._hoverActive = true;
+            };
+
+            this._stop = () => {
```

Comment:
Si on mettait toutes les fonctions genre keyup, keydown tout ça sur this ça marcherait pas ? C'est un peu spéce quand même de redéfinir tout this.stop() à chaque render... ^^'

Reply:
Je trouve ça plus simple d'avoir les fonctions définie localement car elles impliques un scope plus limité.
Si les fonctions se trouvent sur la classe, elles peuvent potentiellement être utilisée autrement.
Ici, on voit très bien à quel moment et pourquoi on bind dans le dom.

Après c'est un détail de style, on pourrait aussi faire:
```js
constructor() {
    this._myHandler = this._myHandler.bind(this);
}
onKeyDown() {
    if (key === '/') this.bindCommandBar();
}
_bindActiveCommanbar() {
    this.document.addEventListener(this._myHandler);
}
_myHandler() {
    //...logic...
    this.stop();
}
_stop() {
    this.document.removeEventListener(this.myHandler)
}
```
Pour moi ça pollue les méthodes de la classes et implique que la fonction pourrait théoriquement être utilisé autre part dans la classe alors qu'en réalité on ne veux pas ça explicitement. Dans notre cas, c'est local après l'activation de la commande.
Et en ça rajoute des lignes où il faut binder les méthodes dans le constructeur.

Reply:
Agreed.